jobbergate-api/tests/apps/job_submissions/test_routers.py

```diff
@@ -2079,3 +2124,498 @@ async def test_job_submissions_agent_active__returns_400_if_token_does_not_carry
+
+    inject_security_header("who@cares.com", permission, client_id="dummy-client")
+    response = await client.put(
+        f"/jobbergate/job-submissions/agent/metrics/{inserted_job_submission_id}",
+        content=encoded_data,
+        headers={"Content-Type": "application/octet-stream"},
+    )
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+    assert response.json() == {"detail": "Invalid data"}
+    mocked_validate_job_metric_upload_input.assert_called_once_with(
+        data, (int, str, int, int, float, float, float, int, float, int, int, int, int, int)
+    )
+
+
+@pytest.mark.parametrize(
+    "permission, num_rows",
+    [
+        (Permissions.ADMIN, 3),
+        (Permissions.JOB_SUBMISSIONS_UPDATE, 8),
+    ],
+)
+async def test_job_submissions_agent_metrics_upload__successful_request(
+    permission,
+    num_rows,
+    fill_job_script_data,
+    fill_job_submission_data,
+    client,
+    inject_security_header,
+    synth_services,
+    synth_session,
+):
+    """
+    Test PUT /job-submissions/agent/metrics/{job_submission_id} returns 204 upon a successful request.
+    """
+    base_job_script = await synth_services.crud.job_script.create(**fill_job_script_data())
+
+    inserted_job_script_id = base_job_script.id
+
+    inserted_submission = await synth_services.crud.job_submission.create(
+        job_script_id=inserted_job_script_id,
+        **fill_job_submission_data(
+            client_id="dummy-client",
+            status=JobSubmissionStatus.SUBMITTED,
+            slurm_job_id=111,
+            slurm_job_state=SlurmJobState.PENDING,
+            slurm_job_info="Fake slurm job info",
+        ),
+    )
+    inserted_job_submission_id = inserted_submission.id
+
+    base_time = int(datetime.now().timestamp())
+    raw_data = generate_job_submission_metric_columns(base_time, num_rows)
+    encoded_data = msgpack.packb(raw_data)
+
+    inject_security_header("who@cares.com", permission, client_id="dummy-client")
+    response = await client.put(
+        f"/jobbergate/job-submissions/agent/metrics/{inserted_job_submission_id}",
+        content=encoded_data,
+        headers={"Content-Type": "application/octet-stream"},
+    )
+    assert response.status_code == status.HTTP_204_NO_CONTENT
+
+    query = select(JobSubmissionMetric).where(
+        JobSubmissionMetric.job_submission_id == inserted_job_submission_id
+    )
+    result = await synth_session.execute(query)
+    scalars = result.scalars()
+    assert all(
+        (
+            scalar.time,
+            scalar.node_host,
+            scalar.step,
+            scalar.task,
+            scalar.cpu_frequency,
+            scalar.cpu_time,
+            scalar.cpu_utilization,
+            scalar.gpu_memory,
+            scalar.gpu_utilization,
+            scalar.page_faults,
+            scalar.memory_rss,
+            scalar.memory_virtual,
+            scalar.disk_read,
+            scalar.disk_write,
```

Comment:
I'm wondering if we should be checking theses values against `raw_data` to make sure the data is making a round trip through the database correctly.

Reply:
but the check is being made against the `raw_data`, check line 2362.

Reply:
Jeez...I completely misread this. Sorry about that.